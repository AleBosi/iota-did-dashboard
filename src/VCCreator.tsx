import React, { useState } from "react";
import { addProofMock } from "./utilsVC";
import VerifyFlag from "./VerifyFlag";
import { Button } from "@/components/ui/button";
import { Input } from "@/components/ui/input";

export default function VCCreator({
  did,
  onCreated,
}: {
  did: string;
  onCreated: (vc: any) => Promise<boolean>;
}) {
  const [vc, setVc] = useState<any>(null);
  const [serial, setSerial] = useState("");
  const [model, setModel] = useState("");
  const [desc, setDesc] = useState("");

  async function createVc() {
    if (!did) {
      alert("DID mancante!");
      return;
    }
    if (!serial || !model || !desc) {
      alert("Compila tutti i campi!");
      return;
    }

    const credential = {
      "@context": ["https://www.w3.org/2018/credentials/v1"],
      type: ["VerifiableCredential", "ProductCertificate"],
      issuer: did,
      issuanceDate: new Date().toISOString(),
      credentialSubject: {
        id: `urn:product:${serial}`,
        model,
        description: desc,
      },
      eventHistory: [],
    };

    const withProof = await addProofMock(credential, did);
    const ok = await onCreated(withProof);
    if (ok) {
      setVc(withProof);
    } else {
      setVc(null);
    }
  }

  function downloadJson(obj: any, filename = "vc.json") {
    const dataStr =
      "data:text/json;charset=utf-8," +
      encodeURIComponent(JSON.stringify(obj, null, 2));
    const a = document.createElement("a");
    a.setAttribute("href", dataStr);
    a.setAttribute("download", filename);
    a.click();
  }

  return (
    <div className="p-6 bg-white rounded-2xl shadow-sm">
      <h1 className="text-2xl font-bold mb-6">Digital Product Passport</h1>

      <h2 className="text-lg font-semibold">DID aziendale</h2>
      <div className="w-full mb-4 bg-gray-100 text-gray-800 font-mono border border-gray-300 rounded-md p-2 break-words">
        {did}
      </div>

      <h2 className="text-lg font-semibold mb-2">
        Crea una Verifiable Credential (VC)
      </h2>

      <Input
        placeholder="Seriale Prodotto"
        value={serial}
        onChange={(e) => setSerial(e.target.value)}
        className="mb-3"
      />
      <Input
        placeholder="Modello"
        value={model}
        onChange={(e) => setModel(e.target.value)}
        className="mb-3"
      />
      <Input
        placeholder="Descrizione"
        value={desc}
        onChange={(e) => setDesc(e.target.value)}
        className="mb-4"
      />

      <Button onClick={createVc} className="mb-6">
        Crea VC
      </Button>

      {vc && (
        <div className="mt-6">
          <h2 className="text-lg font-semibold flex items-center gap-2">
            Step 3: Risultato VC (JSON) <VerifyFlag vc={vc} />
          </h2>
          <pre className="bg-gray-100 p-4 mt-2 rounded-md text-sm text-gray-800 overflow-auto max-h-96">
            {JSON.stringify(vc, null, 2)}
          </pre>
          <Button
            onClick={() => downloadJson(vc)}
            variant="outline"
            className="mt-4"
          >
            Scarica VC come JSON
          </Button>
        </div>
      )}
    </div>
  );
}
