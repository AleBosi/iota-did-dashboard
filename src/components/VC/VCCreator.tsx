import React, { useMemo, useState } from "react";
import { useUser } from "../../contexts/UserContext";
import { useData } from "../../state/DataContext";
import { makeVCIntegrity } from "../../utils/vcIntegrity";

// Tipi minimi compatibili con lo standard VC
type DraftVC = {
  "@context": string[];
  type: string[];
  issuer: string;               // DID dell'emittente (attore corrente)
  issuanceDate: string;         // ISO 8601
  credentialSubject: {
    id: string;                 // DID o identificativo del soggetto
    type?: string | string[];   // opzionale: tipologia subject
    value: string;              // valore/attributo
  };
  // NB: eventHistory NON viene firmata by-design
  eventHistory?: Array<any>;
  // proof sarà aggiunta da makeVCIntegrity
};

export default function VCCreator() {
  const { currentActor } = useUser();
  const { COSTS, spendFromActor, addVC, notify } = useData();

  // form state
  const [subject, setSubject] = useState("");
  const [vcType, setVcType] = useState("");   // es: "Abilitazione"
  const [value, setValue] = useState("");

  const canPublish =
    !!currentActor?.did && subject.trim() !== "" && vcType.trim() !== "" && value.trim() !== "";

  // Prepara un draft VC coerente con lo standard base
  const draft: DraftVC | null = useMemo(() => {
    if (!currentActor?.did || !subject || !vcType || !value) return null;

    return {
      "@context": [
        "https://www.w3.org/2018/credentials/v1",
        // eventuale context custom di dominio
      ],
      type: ["VerifiableCredential", vcType],
      issuer: currentActor.did,
      issuanceDate: new Date().toISOString(),
      credentialSubject: {
        id: subject,
        type: vcType,
        value,
      },
      // opzionale, NON firmata: timeline operativa o append-only log
      eventHistory: [],
    };
  }, [currentActor?.did, subject, vcType, value]);

  const onPublish = async (e: React.FormEvent) => {
    e.preventDefault();
    if (!canPublish || !currentActor || !draft) return;

    try {
      // 1) Addebito crediti (sponsored tx dal wallet master secondo governance)
      await spendFromActor(currentActor.did, COSTS.publishVC);

      // 2) Calcolo proof su payload canonicalizzato (senza proof/eventHistory)
      const withProof = await makeVCIntegrity(draft, { signerDid: currentActor.did });

      // 3) Persistenza VC nel DataContext (fonte unica)
      addVC(withProof);

      // UI feedback + reset form
      notify?.("VC pubblicata con successo ✅");
      setSubject("");
      setVcType("");
      setValue("");
    } catch (err: any) {
      notify?.(`Errore pubblicazione VC: ${err?.message || String(err)}`, "error");
    }
  };

  return (
    <form onSubmit={onPublish} className="flex flex-col gap-3 mb-4">
      <div className="grid grid-cols-1 gap-2">
        <input
          value={subject}
          onChange={(e) => setSubject(e.target.value)}
          placeholder="Subject (DID o nome)"
          className="border px-3 py-2 rounded-xl"
        />
        <input
          value={vcType}
          onChange={(e) => setVcType(e.target.value)}
          placeholder="Tipo credential (es: Abilitazione)"
          className="border px-3 py-2 rounded-xl"
        />
        <input
          value={value}
          onChange={(e) => setValue(e.target.value)}
          placeholder="Valore / Attributo"
          className="border px-3 py-2 rounded-xl"
        />
      </div>

      <div className="flex items-center justify-between">
        <div className="text-xs text-gray-500">
          Costo pubblicazione: <b>{COSTS?.publishVC ?? 1}</b> crediti
        </div>
        <button
          type="submit"
          className="bg-black text-white px-4 py-2 rounded-xl disabled:opacity-40"
          disabled={!canPublish}
          title={!canPublish ? "Compila tutti i campi" : "Pubblica VC"}
        >
          Pubblica VC
        </button>
      </div>

      {/* Anteprima JSON (read-only) utile in fase di debug */}
      {draft && (
        <pre className="mt-2 text-xs bg-gray-50 border rounded-xl p-3 overflow-auto">
          {JSON.stringify(draft, null, 2)}
        </pre>
      )}
    </form>
  );
}
