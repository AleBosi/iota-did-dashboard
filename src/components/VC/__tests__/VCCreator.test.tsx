import { render, screen, fireEvent } from "@testing-library/react";
import VCCreator from "../VCCreator";

describe("VCCreator component", () => {
  it("renderizza i campi e il bottone", () => {
    render(<VCCreator onCreate={jest.fn()} />);
    expect(screen.getByPlaceholderText(/Subject/i)).toBeInTheDocument();
    expect(screen.getByPlaceholderText(/Tipo credential/i)).toBeInTheDocument();
    expect(screen.getByPlaceholderText(/Valore\/Attributo/i)).toBeInTheDocument();
    expect(screen.getByRole("button", { name: /crea vc/i })).toBeInTheDocument();
  });

  it("il bottone è disabilitato se i campi sono vuoti", () => {
    render(<VCCreator onCreate={jest.fn()} />);
    const btn = screen.getByRole("button", { name: /crea vc/i });
    expect(btn).toBeDisabled();
  });

  it("il bottone è abilitato solo se tutti i campi sono compilati", () => {
    render(<VCCreator onCreate={jest.fn()} />);
    fireEvent.change(screen.getByPlaceholderText(/Subject/i), { target: { value: "did:iota:123" } });
    fireEvent.change(screen.getByPlaceholderText(/Tipo credential/i), { target: { value: "Compliance" } });
    fireEvent.change(screen.getByPlaceholderText(/Valore\/Attributo/i), { target: { value: "Certificato OK" } });
    const btn = screen.getByRole("button", { name: /crea vc/i });
    expect(btn).not.toBeDisabled();
  });

  it("chiama onCreate con i valori corretti e resetta i campi", () => {
    const onCreate = jest.fn();
    render(<VCCreator onCreate={onCreate} />);
    fireEvent.change(screen.getByPlaceholderText(/Subject/i), { target: { value: "did:iota:123" } });
    fireEvent.change(screen.getByPlaceholderText(/Tipo credential/i), { target: { value: "Compliance" } });
    fireEvent.change(screen.getByPlaceholderText(/Valore\/Attributo/i), { target: { value: "Certificato OK" } });
    fireEvent.click(screen.getByRole("button", { name: /crea vc/i }));

    expect(onCreate).toHaveBeenCalledWith({
      subject: "did:iota:123",
      type: "Compliance",
      value: "Certificato OK"
    });

    // I campi sono svuotati
    expect(screen.getByPlaceholderText(/Subject/i)).toHaveValue("");
    expect(screen.getByPlaceholderText(/Tipo credential/i)).toHaveValue("");
    expect(screen.getByPlaceholderText(/Valore\/Attributo/i)).toHaveValue("");
  });

  it("NON chiama onCreate se almeno un campo è vuoto", () => {
    const onCreate = jest.fn();
    render(<VCCreator onCreate={onCreate} />);
    // Subject e tipo, ma non valore
    fireEvent.change(screen.getByPlaceholderText(/Subject/i), { target: { value: "did:iota:123" } });
    fireEvent.change(screen.getByPlaceholderText(/Tipo credential/i), { target: { value: "Compliance" } });
    fireEvent.click(screen.getByRole("button", { name: /crea vc/i }));
    expect(onCreate).not.toHaveBeenCalled();
  });

  it("matcha lo snapshot", () => {
    const { asFragment } = render(<VCCreator onCreate={jest.fn()} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
