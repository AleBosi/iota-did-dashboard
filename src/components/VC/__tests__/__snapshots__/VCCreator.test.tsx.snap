// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`VCCreator component matcha lo snapshot 1`] = `
<DocumentFragment>
  <form
    class="flex flex-col gap-2 mb-4"
  >
    <input
      class="border px-2 py-1 rounded"
      placeholder="Subject (DID o nome)"
      value=""
    />
    <input
      class="border px-2 py-1 rounded"
      placeholder="Tipo credential (es: Abilitazione)"
      value=""
    />
    <input
      class="border px-2 py-1 rounded"
      placeholder="Valore/Attributo"
      value=""
    />
    <button
      class="bg-green-600 text-white px-4 py-1 rounded"
      disabled=""
      type="submit"
    >
      Crea VC
    </button>
  </form>
</DocumentFragment>
`;
